### PR TITLE
Kotlin basicsensors

### PR DESCRIPTION
**Pull request by garanj**
Migrates BasicSensors sample to Kotlin

- Migrates to JetPack
- Removes usage of GoogleSignIn#getLastSignedInAccount in favor of getAccountForExtension

### CODE REVIEW

File: BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+ * data sources and to register/unregister listeners to those sources. It also demonstrates how to
+ * authenticate a user with Google Play Services.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions = FitnessOptions.builder().addDataType(DataType.TYPE_LOCATION_SAMPLE).build()
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::findFitnessDataSources to 1
```
Comment (codingjeremy) [inline BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: I would make this a constant so it's clear from the top it's used as a request code.

File: BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+        // When permissions are revoked the app is restarted so onCreate is sufficient to check for
+        // permissions core to the Activity's functionality.
+        checkPermissionsAndRun(::findFitnessDataSources)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (hasRuntimePermissions()) {
```
Comment (codingjeremy) [inline BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: Refactor "hasRuntimePermissions()" to "permissionsApproved()" as it's much easier to understand.

File: BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    }
+
+    /**
+     * Checks that the user is signed in, and if so, executes the specified function. If the user is
+     * not signed in, initiates the sign in flow, specifying the post-sign in function to execute.
+     *
+     * @param signInAction The function to execute if already signed in, or once signed in.
+     */
+    private fun fitSignIn(signInAction: () -> Any) {
+        if (!hasOAuthPermissions()) {
```
Comment (codingjeremy) [inline BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: Do the positive so it's easier to read:

if (hasOAuthPermissions()) {
...

File: BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+            val requestCode = signInActionMap[signInAction]
+            requestCode?.let {
+                GoogleSignIn.requestPermissions(
+                        this,
+                        requestCode,
```
Comment (codingjeremy) [inline BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: it?

File: BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+                        getGoogleAccount(), fitnessOptions)
+            }
+        } else signInAction()
+    }
+
+    private fun hasOAuthPermissions() = GoogleSignIn.hasPermissions(getGoogleAccount(), fitnessOptions)
+
+    private fun getGoogleAccount() = GoogleSignIn.getAccountForExtension(this, fitnessOptions)
```
Comment (codingjeremy) [inline BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: Comment for getAccountForExtension()... not sure if it's different than just an getGoogleAccount() method on GoogleSIgnIn... also, I don't know much about sign in so I might be missing something obvious.

File: BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+    override fun onResume() {
+        super.onResume()
+        // This ensures that if the user denies the permissions then uses Settings to re-enable
+        // them, the app will start working.
+        fitSignIn(::findFitnessDataSources)
```
Comment (codingjeremy) [inline BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: I'm confused by this... aren't you basically doing this in onCreate() so you are doing it twice if they approve in onCreate()?

Comment (garanj) [reply BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt]: Good point! Removed

**Pull request merged**